fish-rust/src/builtins/string/trim.rs

```diff
@@ -0,0 +1,99 @@
+use super::{Arguments, StringError, StringSubCommand};
+use crate::builtins::shared::{io_streams_t, STATUS_CMD_ERROR, STATUS_CMD_OK};
+use crate::ffi::parser_t;
+use crate::wchar::{wstr, L};
+use crate::wchar_ext::WExt;
+use crate::wgetopt::{wopt, woption, woption_argument_t::*};
+
+pub struct Trim<'args> {
+    chars_to_trim: &'args wstr,
+    left: bool,
+    right: bool,
+    quiet: bool,
+}
+
+impl Default for Trim<'_> {
+    fn default() -> Self {
+        Self {
+            // from " \f\n\r\t\v"
+            chars_to_trim: L!(" \x0C\n\r\x09\x0B"),
+            left: false,
+            right: false,
+            quiet: false,
+        }
+    }
+}
+
+impl<'args> StringSubCommand<'args> for Trim<'args> {
+    const LONG_OPTIONS: &'static [woption<'static>] = &[
+        wopt(L!("chars"), required_argument, 'c'),
+        wopt(L!("left"), no_argument, 'l'),
+        wopt(L!("right"), no_argument, 'r'),
+        wopt(L!("quiet"), no_argument, 'q'),
+    ];
+    const SHORT_OPTIONS: &'static wstr = L!(":c:lrq");
+
+    fn parse_opt(&mut self, optarg: Option<&'args wstr>, c: char) -> Result<(), StringError> {
+        match c {
+            'c' => self.chars_to_trim = optarg.unwrap(),
+            'l' => self.left = true,
+            'r' => self.right = true,
+            'q' => self.quiet = true,
+            _ => return Err(StringError::UnknownOption),
+        }
+        return Ok(());
+    }
+
+    fn handle(
+        &mut self,
+        _parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optind: &mut usize,
+        args: &[&wstr],
+    ) -> Option<libc::c_int> {
+        // If neither left or right is specified, we do both.
+        if !self.left && !self.right {
+            self.left = true;
+            self.right = true;
+        }
+
+        let mut ntrim = 0;
+
+        let to_trim_end = |str: &wstr| -> usize {
+            str.chars()
+                .rev()
+                .take_while(|&c| self.chars_to_trim.contains(c))
+                .count()
+        };
+
+        let to_trim_start = |str: &wstr| -> usize {
+            str.chars()
+                .take_while(|&c| self.chars_to_trim.contains(c))
+                .count()
+        };
```

Comment:
> yeah I guess we don't have trim_end_matches() yet

Differences in utility-functions between `widestr` and `str` is a bit frustrating.